Separate button for "Sell Outfits"

UnorderedSigh opened the pull request:
**Feature:** This PR implements an alternative to the features suggested in #8538 and #7055

## Feature Details
The "Sell Outfits" button and "Sell All" button are separate buttons with separate keystrokes. Key `s` with `shift`, `ctrl`, `alt`, etc., only sells commodities and mineables. The keystroke for selling outfits is `shift-x`.

## UI Screenshots
![sell-outfits-button](https://github.com/endless-sky/endless-sky/assets/116329264/3c6bf0e3-b203-4f39-8668-76eb602f5c83)

## Usage Examples
Don't sell your outfits by accident.

## Testing Done
Put outfits and commodities in my cargo bay and wandered from planet to planet seeing if things work as intended.

### Automated Tests Added
Can't think of any.

## Performance Impact
N/A

UnorderedSigh commented:
Pinging @williaji who requested both of the alternatives to this feature.

UnorderedSigh commented:
The `Sell All` and `Buy All` buttons are slightly confusing now. You'd think  `All` means all, but it just means `Sell/Buy All Commodities`. Endless Sky buttons can't get any wider than this. Does anyone have a suggestion on how to make the differences between the buttons more clear?

williaji commented:
> The `Sell All` and `Buy All` buttons are slightly confusing now. You'd think `All` means all, but it just means `Sell/Buy All Commodities`. 

Within the context of the full screen, the 'Sell Outfits' button makes a distinction between commodities and outfits even with the current buttons being somewhat vague. I'm not a new player, so I can't say that there wouldn't be some risk of confusion, but I think most folks would work it out fairly quickly. I think it also helps that 'Buy/Sell All' buttons are located under the relevant portions of the commodities display; it provides visual context. I doubt we are going to find a better wording given the space available.

UnorderedSigh commented:
> I think most folks would work it out fairly quickly

If you have to work out a new interface, then it is poorly designed and needs to be changed. This one isn't just new to new players. Nobody has used it before.  The entire community will have to "work it out," and I don't want them to have to do that.

lumbar527 commented:
What if after you clicked `Sell All` it turned into `Sell Outfits`?

UnorderedSigh commented:
> What if after you clicked `Sell All` it turned into `Sell Outfits`?

I'm not sure if you're being facetious. That is the exact feature we want to remove.

lumbar527 commented:
Sorry, I just thought it would be a little clearer if the text changed. I guess that wouldn't make as much difference as I thought it could.

UnorderedSigh commented:
> Sorry, I just thought it would be a little clearer if the text changed. I guess that wouldn't make as much difference as I thought it could.

It does, indeed. It makes people accidentally sell their outfits. You are correct though, that it makes the button meaning abundantly clear. I need a way to make the button's meaning clear without changing its meaning after you press it.

UnorderedSigh commented:
> The one edge case that I've noticed in testing is that special commodities (i.e. mineables) are included in the 'Sell All.' They are called out in the text manifest in the left corner and will be sold with the rest of the commodities. I don't see this as a bug per se, but it is another case where we are overloading trade categories into a single function. Practically, this is unlikely to be a major problem. The worst case I can see is that someone has a mining job and accidentally sells the cargo. Given that there are few mining jobs and asteroids respawn, this seems primarily an inconvenience.

It would be easy to add a `Sell Other` button.

UnorderedSigh commented:
Or `Sell Specials`, since they're explicitly labeled as "*special* commodities."

williaji commented:
> Or `Sell Specials`, since they're explicitly labeled as "_special_ commodities."

A separate button would break out all the categories which makes trade commodities clearer.

Would it be possible to include the 'Buy/Sell All' buttons within the commodity frame? If the commodity table had an irregular border that included the 'Buy/Sell All' buttons it would make it clear that they are associated. The other two, being outside the border would be seen as separate.

UnorderedSigh commented:
> Would it be possible to include the 'Buy/Sell All' buttons within the commodity frame? If the commodity table had an irregular border that included the 'Buy/Sell All' buttons it would make it clear that they are associated. The other two, being outside the border would be seen as separate.

I'm not sure what you're asking for. Could you please post an image explaining what you want?

The easiest way is to put rough boxes on an edited version of an image in my PR description. It doesn't need to be elegant. You can hide it behind a `<details><summary> The image </summary> ... </details>` block if you wish.

williaji commented:
![Screenshot_20230607_190749](https://github.com/endless-sky/endless-sky/assets/70952724/fc461ceb-62db-493e-931e-ef3e33406c2a)

UnorderedSigh commented:
Oh! Okay, It is obvious to me now. I will ponder how to make an outline.

UnorderedSigh commented:
I have added the [Sell Specials] button. The keystroke is shift-m.

UnorderedSigh commented:
@williaji - I put an outline around the buttons. How does it look?

![add-trading-panel-extension](https://github.com/endless-sky/endless-sky/assets/116329264/e532517a-e692-4cf4-b45e-a1b314740853)

CodeKatta commented:
> I put an outline around the buttons. How does it look?

Although I haven't yet managed to sell outfits in cargo by accident, I think it's a good safeguard against doing so.
And just recently I had to temporarily move some mineables from cargo to storage and back again so I didn't automatically sell them.

An other alternative would be to include an [all]-button after every commodity: [buy]     [sell] [all]
Don't know how much that would interfere with the current setup but the changes so far look like a useful improvement and quite intuitive, especially if there is an additional explanation.

UnorderedSigh commented:
> An other alternative would be to include an [all]-button after every commodity: [buy]     [sell] [all]
> Don't know how much that would interfere with the current setup but the changes so far look like a useful improvement and quite intuitive, especially if there is an additional explanation.

Redoing the commodity trading area is outside the scope of this PR. If you want it to happen, please open an Issue with specific details. A mock-up user interface picture would help.

williaji commented:
This is working very well for me. It clearly segments the various trade types and requires you to make a choice to sell them. It's easy to both carry outfits and trade along the way. Each set of controls is clearly marked and the revised border collects the commodities block. "Specials" is vague, but that's what the game uses as we doubled up on the term "commodities." We might consider "Minerals" but I don't know if there is anything else that could fall into the "special" category. I'm delighted with this.

UnorderedSigh commented:
I would like a better option than `[Sell Specials]`. From the point of view of a new player, and the code, `[Sell Minables]` makes the most sense. However, the same category is what you would use for other non-standard commodities.

UnorderedSigh commented:
> I thought I first mention it here since an option to sell all of one type of commodity with a single click could also be a solution to the problem of selling things you do not want to sell (of course, one could also use <shift|ctrl|alt> but having a simple button for that is probably more convenient for most people).

On the subject of modifying the trading table, I'd prefer to replace it entirely. There should be a better way to present that information.

UnorderedSigh commented:
@williaji - I just implemented the two features you asked for. 

1. Confirmation dialog if you press `[Sell Outfits]`. You can turn this off in the settings.
2. A setting to disable that button on planets without an outfitter.

By default, you are allowed to sell outfits without an outfitter, and you get a confirmation dialog.

williaji commented:
Having settings for these is a nice touch, as is the tutorial  panel. This is a very thorough PR. Thank you.

I think the confirmation dialog box would be better without the pointer to the setting. I found it redundant to be reminded of a setting every time I sold outfits. You might mention the setting in the tutorial, but I would pull it from the dialog itself.

Since the purpose of this dialog is to prevent accidental sales, the default to should be 'Cancel' rather than 'OK'.

williaji commented:
One other thought on the outfit sale confirmation. You might have it list the outfits to be sold the way the ship sale confirmation does. I will usually go to the player info screen and list the cargo before I sell outfits and that would save a step.

UnorderedSigh commented:
> One other thought on the outfit sale confirmation. You might have it list the outfits to be sold the way the ship sale confirmation does. I will usually go to the player info screen and list the cargo before I sell outfits and that would save a step.

That might be *extremely* long.

UnorderedSigh commented:
The deed is done.

![confirm-sell-mineable](https://github.com/endless-sky/endless-sky/assets/116329264/7da5873b-6b05-4c67-b25e-a943bb7cb241)


![sell-outfits-list](https://github.com/endless-sky/endless-sky/assets/116329264/1d3354c4-762a-4790-ab46-71f6bd627b0d)

UnorderedSigh commented:
The outfit list is sorted in order from most valuable (value * count) to least.

quyykk commented:
I'm not really a fan of the preferences, because not everything should be a preference. So I'll propose the following: Show the confirmation dialog when clicking only on planets without an outfitter.

UnorderedSigh commented:
> I'm not really a fan of the preferences, because not everything should be a preference. So I'll propose the following: Show the confirmation dialog when clicking only on planets without an outfitter.

What about minables?

UnorderedSigh commented:
I *definitely* would want to be able to disable the dialog. It'll get annoying pretty quickly.

quyykk commented:
> What about minables?

Not sure I understand?

> I definitely would want to be able to disable the dialog. It'll get annoying pretty quickly.

How often are you selling outfits on planets without an outfitter?

UnorderedSigh commented:
>  What about minables?

I added a confirmation dialog for minables. It lists the minables you're going to sell and asks if you're certain.

> How often are you selling outfits on planets without an outfitter?

I'm not always near an outfitter when I'm ~~farming the innocent~~. Especially at Durax. The outfitter is two jumps farther.

UnorderedSigh commented:
I definitely don't want to remove the settings to disable the confirmations. I will frequently want them on or off:

1. Farming outfits for sale or mining asteroids for sale. Turn off the warnings so you don't have to press extra buttons every time you land.
2. Pillaging to upgrade your ship, or mining asteroids for missions. Turn on the warnings so you don't accidentally sell what you need.

williaji commented:
I've been running with this today and having the listing in the confirmation is positive for me. I've moved outfits around and haven't lost anything. I hope this gets added because I'm spoiled now and wouldn't want to play without it.

lumbar527 commented:
Might be a little late, but I found a comment referring to this sort of thing: https://github.com/endless-sky/endless-sky/discussions/7525#discussioncomment-4190806

UnorderedSigh commented:
> Might be a little late, but I found a comment referring to this sort of thing: https://github.com/endless-sky/endless-sky/discussions/7525#discussioncomment-4190806

Thank you. I have added that to the PR description. It is certainly not too late; it'll likely be months, if not longer, before this PR is merged.

UnorderedSigh commented:
Last tested and known to work Sun, 27 Aug 2023 15:51:53 UTC

UnorderedSigh commented:
Last tested and known to work Thu, 23 Nov 2023 01:00:35 UTC

TomGoodIdea commented:
@UnorderedSigh Could you please resolve merge conflicts?

UnorderedSigh commented:
Please add "waiting for OP" to this to remind me that it is on the list of PRs you need me to update.

nickshanks commented:
I had no idea this existed and have made an almost identical but much simpler version.

* I called Specials "Minerals" since it specifically applyes to things that are have `"minable" > 0.0` and Minerals is the term used in the Outfitters for these items.
* I did not make a custom graphic for it.
* I did not add, remove, or change any dialogs.

I prefer to do lots of small PRs that are easy to approve and merge and build on, than one large PR that hangs around for over a year and needs frequent merge conflict resolution.

If mine gets merged in lieu of this I would be happy to migrate over the missing features from this into another PR.

UnorderedSigh commented:
> I had no idea this existed and have made an almost identical but much simpler version.

I've already been informed that nearly all of my PRs will never be merged. I suspect this is purely politics. That means, practically speaking, for any of these features to be merged, someone else will have to re-implement them.  Continuing to maintain them is out of a sense of obligation.

UnorderedSigh commented:
> I prefer to do lots of small PRs that are easy to approve and merge and build on, than one large PR that hangs around for over a year and needs frequent merge conflict resolution.

I've implemented many complex features that people have been requesting for years. If I split them up into little PRs, I'd have another fifty PRs, and that would be depressing to the reviewers.

tibetiroka commented:
> I've already been informed that nearly all of my PRs will never be merged. I suspect this is purely politics. That means, practically speaking, for any of these features to be merged, someone else will have to re-implement them. Continuing to maintain them is out of a sense of obligation.

This is not my understanding at all. In fact, I have recently milestoned four of your PRs for the next unstable cycle.

I know that we take too long to get to PRs, especially larger code PRs. It's nothing personal, at least not form me - we simply don't have enough manpower. Look at Peter's efforts at getting formation flying into the game, for instance. It's been ongoing for five years, and he's one of the devs.

Per these two PRs, I actually really like the popup confirmation message you implemented, which seems to be the main difference in the UI/UX design of these PRs (though I haven't seen screenshots from the new PR).

UnorderedSigh commented:
> This is not my understanding at all. In fact, I have recently milestoned four of your PRs for the next unstable cycle.

Endless Sky reviewers do a fantastic job. I have no complaints on that account.

UnorderedSigh commented:
> especially larger code PRs

Your group has too few code managers with too little time. No amount of determination nor political stability can solve that. You need more volunteers for code manager positions.

And seriously, a few hundred lines is not a large code PR. In most of the projects I work on, my contributions are 1000 lines minimum. One of them recently was 12,000 lines of new code. They rarely wait in the queue longer than two months. But those projects have an abundance of staff.

UnorderedSigh commented:
This is getting wildly off-topic and needlessly political. I encourage moderators to clean up the thread if they feel that is justified. You have my apologies.

TomGoodIdea commented:
So, what's the status of this PR?

UnorderedSigh commented:
Last tested and known to work Tue, 27 Aug 2024 00:11:35 UTC

TomGoodIdea commented on source/Preferences.cpp:
```diff
@@ -222,6 +225,12 @@ void Preferences::Load()
 			previousSaveCount = max<int>(3, node.Value(1));
 		else if(node.Token(0) == "alt-mouse turning")
 			settings["Control ship with mouse"] = (node.Size() == 1 || node.Value(1));
+		else if(node.Token(0) == "Confirm 'Sell Outfits' button")
+			settings["Confirm 'Sell Outfits' button"] = (node.Size() == 1 || node.Value(1));
+		else if(node.Token(0) == "Confirm 'Sell Specials' button")
+			settings["Confirm 'Sell Specials' button"] = (node.Size() == 1 || node.Value(1));
+		else if(node.Token(0) == "Allow 'Sell Outfits' with no outfitter")
+			settings["'Sell Outfits' without outfitter"] = (node.Size() == 1 || node.Value(1));
```
Why does this need to be explicitly stated? Won't the default handling work?

UnorderedSigh replied:
Apparently the default handling does work.

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -123,31 +129,21 @@ void TradingPanel::Draw()
-	sellOutfits = false;
-	if(player.Cargo().HasOutfits() || missionCargo)
+	double minableCargo = player.Cargo().MinablesSizePrecise();
+	double outfitCargo = max(0.0, player.Cargo().OutfitsSizePrecise() - minableCargo);
```
```suggestion
	double outfitCargo = max(0., player.Cargo().OutfitsSizePrecise() - minableCargo);
```

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -337,3 +336,79 @@ void TradingPanel::Buy(int64_t amount)
+	for(const auto &it : player.Cargo().Outfits())
+	{
+		if(sellMinable != static_cast<bool>(it.first->Get("minable")))
+			continue;
```
```suggestion
			continue;
		if(!it.second)
			continue;
```
either add it here too or remove from the other place below

UnorderedSigh replied:
That is special handling for when the `Outfit*` in the Outfits map is nullptr. I'm not sure that's possible.

TomGoodIdea replied:
idk, I think I saw it once in the existing code

UnorderedSigh replied:
The advantage of selling the nullptrs is that it cleans them out of the map.

The disadvantage is that I don't understand why they'd be in there in the first place, so I don't know how to correctly handle them, if they exist at all.

TomGoodIdea replied:
It's not for nullptrs, it's when you have 0 outifts. Looks like our current code isn't consistent on this matter, but CargoHold methods used for transferring outfits don't clear empty entries, so this check saves some unnecessary calls to depreciation and other stuff.

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -337,3 +336,79 @@ void TradingPanel::Buy(int64_t amount)
+
+string TradingPanel::OutfitSalesMessage(bool sellMinable, size_t displayLimit)
+{
+	struct OutfitInfo
+	{
```
```suggestion
	struct OutfitInfo {
```

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -337,3 +336,79 @@ void TradingPanel::Buy(int64_t amount)
+{
+	struct OutfitInfo
+	{
+		std::string name;
```
```suggestion
		string name;
```

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -337,3 +336,79 @@ void TradingPanel::Buy(int64_t amount)
+		profit += value;
+		tonsSold += static_cast<int>(it.second * it.first->Mass());
+		// Store a description of the count & item, followed by its value.
+		outfitValue.push_back({ "", it.second, value });
```
```suggestion
		outfitValue.emplace_back("", it.second, value);
```
this should work as we use C++20 now

UnorderedSigh replied:
The mac and windows compilers don't like that change.

TomGoodIdea replied:
According to the [implementation matrix](https://en.cppreference.com/w/cpp/20) ("Parenthesized initialization of aggregates"), Windows _should_ support this, but yeah, looks like Apple doesn't like it.

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -337,3 +336,79 @@ void TradingPanel::Buy(int64_t amount)
+		<< Format::CargoString(tonsSold, sellMinable ? "of special commodities" : "of outfits")
+		<< " for " << Format::CreditString(profit) << '?' << endl;
+
+	// Sort by decreasing value
```
```suggestion
	// Sort by decreasing value.
```

TomGoodIdea commented on data/_ui/help.txt:
```diff
@@ -172,6 +172,7 @@ help "stranded"
 help "trading"
 	`This is the trading panel. Earn money by buying commodities at a low price in one system, and selling at a higher price elsewhere. To view your map of commodity prices in other systems, press <View star map>. To buy or sell, click on [buy] or [sell], or select a line with the up and down arrows and press "+" or "-" (or Enter and Delete).`
 	`You can buy 5 tons at once by holding down Shift, 20 by holding down Control, 500 by holding down Alt, or larger amounts at a time by holding down two keys at once, or even all three.`
+	`Sell other items in your cargo hold with the buttons to the left. "Sell Outfits" sells any outfitter items in your cargo hold. "Sell Specials" sells materials you mine from asteroids.`
```
Minables are "outfitter items", too.
```suggestion
	`Sell other items in your cargo hold with the buttons to the left. "Sell Specials" sells materials you mine from asteroids. "Sell Outfits" sells any other outfitter items in your cargo hold.`
```

UnorderedSigh replied:
Special commodities aren't necessarily only mineables. For now though, others are a niche case.

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -123,31 +129,21 @@ void TradingPanel::Draw()
-	sellOutfits = false;
-	if(player.Cargo().HasOutfits() || missionCargo)
+	double minableCargo = player.Cargo().MinablesSizePrecise();
+	double outfitCargo = max(0., player.Cargo().OutfitsSizePrecise() - minableCargo);
```
If you want to avoid floating-point errors, maybe just make a Cargo member function that calculates only non-minables.

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -209,10 +204,14 @@ void TradingPanel::Draw()
+	if(canSellOutfits)
 		info.SetCondition("can sell outfits");
-	else if(player.Cargo().HasOutfits() || canSell)
+	if(minableCargo)
```
For some reason, I can do this:
![zero](https://github.com/user-attachments/assets/6162cb5c-5247-4be5-9513-3bde12567c4e)
And I don't think it's a rounding error, because it can happen immediately after reloading the save.

UnorderedSigh replied:
Can you give me a save file?

xobes replied:
Happens any time you don't have anything to sell and you click Sell Specials.

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -337,3 +336,80 @@ void TradingPanel::Buy(int64_t amount)
+	};
+	vector<OutfitInfo> outfitValue;
+	double tonsSold = 0;
+	int profit = 0;
```
```suggestion
	int64_t profit = 0;
```

TomGoodIdea commented on source/TradingPanel.cpp:
```diff
@@ -337,3 +336,80 @@ void TradingPanel::Buy(int64_t amount)
+			continue;
+		int64_t value = player.FleetDepreciation().Value(it.first, day, it.second);
+		profit += value;
+		tonsSold += static_cast<int>(it.second * it.first->Mass());
```
Why?
```suggestion
		tonsSold += it.second * it.first->Mass();
```

TomGoodIdea reviewed (commented):
These settings are pretty self-explanatory, but I think there's a requirement that every setting should have a tooltip (in `data/_ui/tooltips.txt`).

UnorderedSigh commented:
> Quick question. It the drop-down containing the "Buy All" and "Sell All" buttons required?

Yes. Without this, the "all" is ambiguous. It could refer to all items you own, or all items in your cargo bay, or all commodities. By putting it in the same user interface element, it is clear that the buttons refer only to what is in that user interface element.

TheGiraffe3 commented:
This removes the hotkeys for Sell/Buy All as well, correct? Might be nice to keep those if possible.

xobes commented on source/Preferences.cpp:
```diff
@@ -172,6 +172,9 @@ void Preferences::Load()
 	settings["Target asteroid based on"] = true;
+	settings["Confirm 'Sell Outfits' button"] = true;
+	settings["Confirm 'Sell Specials' button"] = true;
+	settings["'Sell Outfits' without outfitter"] = true;
```
I was going to say "I think the default for the "'Sell Outfits' without outfitter" option should be false", but then I realized you can pick up random crap floating through space and have no way to get rid of it...

I really like the idea of a dialog to confirm outfit sales.

xobes commented:
Is there any way for me to help see this through?  @UnorderedSigh do you need help finishing this?

xobes commented:
Also, are "Specials" explicitly "Minables"?
